db/compaction/compaction_job.cc

```diff
@@ -353,35 +356,75 @@ void CompactionJob::GenSubcompactionBoundaries() {
+                                      max_subcompactions, max_output_files});
+  if (c->immutable_options()->compaction_pri == kRoundRobin) {
+    // The number of files in the start level cannot be larger than
+    // ranges.size() since ranges also contains the number of files from other
+    // levels Note that mutex has been locked, so it is safe to access
+    // bg_compaction_scheduled and update it
+    subcompactions = std::min(
+        {static_cast<uint64_t>(c->num_input_files(0)), max_output_files,
+         static_cast<uint64_t>(available_compactions_ + 1)});
+    // We only reserve threads if the required subcompactions is more than the
+    // number of allowed subcompactions
+    if (max_subcompactions < subcompactions) {
+      int available_threads = env_->ReserveThreads(
+          (int)(subcompactions - max_subcompactions), thread_pri_);
+      // The reserved threads might be less than the expected one, we have to
+      // reduce the subcompactions to enforce the thread limit
+      subcompactions =
+          static_cast<uint64_t>(available_threads) + max_subcompactions;
+      if (bg_compaction_scheduled_ != nullptr) {
+        *bg_compaction_scheduled_ += available_threads;
+      }
+      reserved_threads_ = available_threads;
+    }
+  }
 
   if (subcompactions > 1) {
     double mean = sum * 1.0 / subcompactions;
     // Greedily add ranges to the subcompaction until the sum of the ranges'
     // sizes becomes >= the expected mean size of a subcompaction
     sum = 0;
+    uint64_t num_actual_subcompactions = 0U;
     for (size_t i = 0; i + 1 < ranges.size(); i++) {
       sum += ranges[i].size;
-      if (subcompactions == 1) {
+      if (num_actual_subcompactions + 1 == subcompactions) {
+        num_actual_subcompactions++;
         // If there's only one left to schedule then it goes to the end so no
         // need to put an end boundary
         continue;
       }
       if (sum >= mean) {
         boundaries_.emplace_back(ExtractUserKey(ranges[i].range.limit));
-        subcompactions--;
+        num_actual_subcompactions++;
         sum = 0;
       }
     }
+    // The actual number of subcompactions might be less than the expected
+    // subcompaction, we need to release some threads to avoid wasting the
+    // resource
+    if (reserved_threads_ > 0 && num_actual_subcompactions < subcompactions) {
+      // We cannot release threads more than what we reserved before
+      int released_threads = env_->ReleaseThreads(
+          std::min((int)(subcompactions - num_actual_subcompactions),
+                   reserved_threads_),
+          thread_pri_);
+      // Update the number of reserved threads and the number of background
+      // scheduled compactions for this compaction job
+      reserved_threads_ -= released_threads;
+      if (bg_compaction_scheduled_ != nullptr) {
+        *bg_compaction_scheduled_ -= released_threads;
+      }
+    }
   }
```

Comment:
I doubt whether this complexity is needed for the following reason:
(1) According to our offline chat, the rationale behind `num_actual_subcompactions` is `subcompactions` might not be used up in the loop `for (size_t i = 0; i + 1 < ranges.size(); i++){}` hence we might end up reserving more threads as it's based on the `subcompactions`.

Do you have an actual example or test case that shows `num_actual_subcompactions != subcompactions` or is it just cautious suspicion?

(2) What's below `// Greedily add ranges to the subcompaction until the sum of the ranges'
    // sizes becomes >= the expected mean size of a subcompaction` seems to be averaging "range load" of each sub-compaction. Visually speaking, it's doing below:

Input: 
- Ranges with size indicated as the digit: [--5---][-2-][--3-][---7----][--4-]
- #subcompactions = 3
- mean = (5 + 2 + 3 + 7 + 4)/3 = 7

Output:
- Subcompactions (each indicated as a pair of parentheses) after being averaged on "range load": 
{[--5---][-2-]}, {[--3-][---7----]}, {[--4-]}


I haven't rigorously prove `#subcompactions == num_actual_subcompactions` always holds but it feels true just looking at the visual ideas.

Reply:
Actually here I just reused the existing logic to determine the number of actual subcompactions. The final subcompaction is determined on the size of `boundaries_` instead of `subcompactions` variable. I'm not sure if there exists such a unit case that covers the case you requested.

Reply:
Since the end condition is `i + 1 < ranges.size()`, I believe that if the size is unbalanced where the last range is too large that the number of real subcompactions can be less than the allowed one.

Reply:
Right - that's a good case!